pkg/controller/webhooks/deletionprotection.go

```diff
@@ -0,0 +1,123 @@
+package webhooks
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/vshn/appcat/v4/pkg/comp-functions/runtime"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	k8sruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	errNoOwnerRefAnnotation = fmt.Errorf("ownerReference annotation not specified")
+)
+
+type compositeInfo struct {
+	Exists bool
+	Name   string
+}
+
+// DeletionProtectionInfo provides information about the composite's deletion
+// protection state.
+type DeletionProtectionInfo interface {
+	IsDeletionProtected() bool
+}
+
+// checkManagedObject will find the highest composite for any object that is deployed via Crossplane or provider-kubernetes.
+// For example: XObjectBucket, Release, Namespace, etc.
+// Anything that either contains an owner reference to a composite or an owner annotation.
+func checkManagedObject(ctx context.Context, obj client.Object, c client.Client, l logr.Logger) (compositeInfo, error) {
```

Comment:
This function can halt deletion of namespaces and PVCs in the whole cluster if I understood it right, please test it throughly.

Reply:
Ongoing ;)